File: Java/Daily/Medium/FindMinTimeBrewPotions.java
// 3494
class Solution {
    public long minTime(int[] skill, int[] mana) {
        int n = skill.length;   // number of wizards
        int m = mana.length;    // number of potions

        // Compute prefix sum of wizard skills
        long[] prefixSkill = new long[n];
        for (int i = 1; i < n; i++) {
            prefixSkill[i] = prefixSkill[i - 1] + skill[i];
        }

        // Time for the first potion at the first wizard
        long potionStartTime = (long) skill[0] * mana[0];

        // Loop over remaining potions
        for (int j = 1; j < m; j++) {
            long maxDiff = (long) skill[0] * mana[j]; // includes first wizard

            // Check bottleneck for wizards 1 to n-1
            for (int i = 1; i < n; i++) {
                long currDiff = (long) mana[j - 1] * prefixSkill[i] - (long) mana[j] * prefixSkill[i - 1];
                if (currDiff > maxDiff) {
                    maxDiff = currDiff;
                }
            }

            // Add bottleneck delay for this potion
            potionStartTime += maxDiff;
        }

        // Add time for last potion to pass through all wizards
        long totalTime = potionStartTime + (long) mana[m - 1] * prefixSkill[n - 1];
        return totalTime;
    }
}

File: Daily/FindMinTimeBrewPotions.py
from typing import List
# 3494
class Solution:
    def minTime(self, skill: List[int], mana: List[int]) -> int:

        m, n = len(mana), len(skill)
        prefix_skill = [0] * n

        for i in range(1, n):
            prefix_skill[i] = prefix_skill[i - 1] + skill[i]
        # prefix_skill[i] denotes total skill of wizards 0 to i
        
        potion_start_time = skill[0] * mana[0]
        for j in range(1, m):
            max_diff = skill[0] * mana[j]
            for i in range(1, n):
                """
                Bottleneck is caused by largest wait time, wait time is found after two conditions:
                    - current wizard should have finished the previous potion
                    - current potion should be brewed the previous wizard
                Obviously it takes longer for the current wizard to reach the current potion,
                that'i why the difference is calculated shown below
                """
                curr_diff = mana[j - 1] * prefix_skill[i] - mana[j] * prefix_skill[i - 1]
                if curr_diff > max_diff:
                    max_diff = curr_diff
            
            potion_start_time += max_diff
        # We have the start time of the last potion so we just add product of last potion and skill of all wizards.
        total = potion_start_time + mana[-1] * prefix_skill[-1]
        return total